File: crates/nge2-formats/src/hgpt.rs
use crate::{align, u16_le, u32_le, FormatError};
use serde::Serialize;
use specta::Type;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct HgptImage {
    pub width: u32,
    pub height: u32,
    pub pixel_format: HgptPixelFormat,
    pub divisions: Vec<HgptDivision>,
    #[serde(skip)]
    #[specta(skip)]
    pub rgba: Vec<u8>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct HgptDivision {
    pub x: u16,
    pub y: u16,
    pub width: u16,
    pub height: u16,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Type)]
pub enum HgptPixelFormat {
    Indexed4,
    Indexed8,
    Rgba8888,
}

impl HgptImage {
    pub fn parse(data: &[u8]) -> Result<Self, FormatError> {
        if data.get(..4) != Some(b"HGPT") {
            return Err(invalid("missing HGPT signature"));
        }
        let pp_offset = usize::from(u16_le(data, 4)?);
        if pp_offset < 16 || pp_offset + 48 > data.len() {
            return Err(invalid("PP offset is outside the file"));
        }
        let has_extended_header = u16_le(data, 6)?;
        if has_extended_header > 1 {
            return Err(invalid("extended header flag must be zero or one"));
        }
        let division_count = usize::from(u16_le(data, 8)?);
        let mut divisions = Vec::with_capacity(division_count);
        if has_extended_header == 1 {
            if u16_le(data, 16)? as usize != division_count {
                return Err(invalid("division count does not match its repeat"));
            }
            let mut cursor = 28;
            for _ in 0..division_count {
                divisions.push(HgptDivision {
                    x: u16_le(data, cursor)?,
                    y: u16_le(data, cursor + 2)?,
                    width: u16_le(data, cursor + 4)?,
                    height: u16_le(data, cursor + 6)?,
                });
                cursor += 8;
            }
            if cursor > pp_offset {
                return Err(invalid("division table overlaps PP data"));
            }
        }

        let pp_header = u32_le(data, pp_offset)?;
        if pp_header & 0xffff != 0x7070 {
            return Err(invalid("missing PP header"));
        }
        let format_code = (pp_header >> 16) as u16;
        let (pixel_format, tile_width, bytes_per_pixel) = match format_code {
            0x14 => (HgptPixelFormat::Indexed4, 32usize, 0.5f32),
            0x13 => (HgptPixelFormat::Indexed8, 16usize, 1.0f32),
            0x8800 => (HgptPixelFormat::Rgba8888, 4usize, 4.0f32),
            _ => {
                return Err(FormatError::Unsupported {
                    format: "HGPT",
                    message: format!("pixel format 0x{format_code:04X}"),
                })
            }
        };
        let width = usize::from(u16_le(data, pp_offset + 4)?);
        let height = usize::from(u16_le(data, pp_offset + 6)?);
        if width == 0 || height == 0 || width.saturating_mul(height) > 32_000_000 {
            return Err(invalid("invalid image dimensions"));
        }
        let ppd_offset = pp_offset + 16;
        let ppd_header = u32_le(data, ppd_offset)?;
        if ppd_header & 0x00ff_ffff != 0x0064_7070
            || (ppd_header >> 24) as u8 != format_code as u8
        {
            return Err(invalid("missing or mismatched PPD header"));
        }
        if usize::from(u16_le(data, ppd_offset + 4)?) != width
            || usize::from(u16_le(data, ppd_offset + 6)?) != height
        {
            return Err(invalid("PP and PPD dimensions do not match"));
        }
        let storage_width = align(width, tile_width);
        let storage_height = align(height, 8);
        let pixel_count = storage_width
            .checked_mul(storage_height)
            .ok_or_else(|| invalid("storage dimensions overflow"))?;
        let data_bytes = (pixel_count as f32 * bytes_per_pixel) as usize;
        let pixel_offset = ppd_offset + 32;
        let tiled = data
            .get(pixel_offset..pixel_offset + data_bytes)
            .ok_or(FormatError::UnexpectedEof(pixel_offset))?;

        let palette = if pixel_format == HgptPixelFormat::Rgba8888 {
            Vec::new()
        } else {
            parse_palette(data, pixel_offset + data_bytes)?
        };
        let expected_palette = match pixel_format {
            HgptPixelFormat::Indexed4 => 16,
            HgptPixelFormat::Indexed8 => 256,
            HgptPixelFormat::Rgba8888 => 0,
        };
        if palette.len() != expected_palette {
            return Err(invalid(&format!(
                "expected {expected_palette} palette colors, found {}",
                palette.len()
            )));
        }

        let mut rgba = vec![0u8; width * height * 4];
        for y in 0..height {
            for x in 0..width {
                let tiled_index = tile_index(x, y, storage_width, tile_width);
                let color = match pixel_format {
                    HgptPixelFormat::Indexed4 => {
                        let byte = tiled[tiled_index / 2];
                        let index = if tiled_index % 2 == 0 { byte & 0x0f } else { byte >> 4 };
                        palette[index as usize]
                    }
                    HgptPixelFormat::Indexed8 => palette[tiled[tiled_index] as usize],
                    HgptPixelFormat::Rgba8888 => {
                        let source = tiled_index * 4;
                        [
                            tiled[source],
                            tiled[source + 1],
                            tiled[source + 2],
                            decode_alpha(tiled[source + 3]),
                        ]
                    }
                };
                let destination = (y * width + x) * 4;
                rgba[destination..destination + 4].copy_from_slice(&color);
            }
        }

        Ok(Self {
            width: width as u32,
            height: height as u32,
            pixel_format,
            divisions,
            rgba,
        })
    }

    pub fn encode_png(&self) -> Result<Vec<u8>, FormatError> {
        let mut output = Vec::new();
        {
            let mut encoder = png::Encoder::new(&mut output, self.width, self.height);
            encoder.set_color(png::ColorType::Rgba);
            encoder.set_depth(png::BitDepth::Eight);
            let mut writer = encoder
                .write_header()
                .map_err(|error| FormatError::Png(error.to_string()))?;
            writer
                .write_image_data(&self.rgba)
                .map_err(|error| FormatError::Png(error.to_string()))?;
        }
        Ok(output)
    }
}

fn parse_palette(data: &[u8], offset: usize) -> Result<Vec<[u8; 4]>, FormatError> {
    if u32_le(data, offset)? != 0x0063_7070 {
        return Err(invalid("missing PPC palette header"));
    }
    let count = usize::from(u16_le(data, offset + 6)?) * 8;
    let colors_offset = offset + 16;
    let raw = data
        .get(colors_offset..colors_offset + count * 4)
        .ok_or(FormatError::UnexpectedEof(colors_offset))?;
    Ok(raw
        .chunks_exact(4)
        .map(|color| [color[0], color[1], color[2], decode_alpha(color[3])])
        .collect())
}

fn tile_index(x: usize, y: usize, storage_width: usize, tile_width: usize) -> usize {
    let tile_height = 8;
    let tile_size = tile_width * tile_height;
    let tile_row = tile_size * (storage_width / tile_width);
    (y / tile_height) * tile_row
        + (x / tile_width) * tile_size
        + (y % tile_height) * tile_width
        + (x % tile_width)
}

fn decode_alpha(value: u8) -> u8 {
    value.saturating_mul(2)
}

fn invalid(message: &str) -> FormatError {
    FormatError::Invalid {
        format: "HGPT",
        message: message.into(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn rgba_fixture() -> Vec<u8> {
        let width = 4u16;
        let height = 8u16;
        let mut data = Vec::new();
        data.extend_from_slice(b"HGPT");
        data.extend_from_slice(&16u16.to_le_bytes());
        data.extend_from_slice(&0u16.to_le_bytes());
        data.extend_from_slice(&0u16.to_le_bytes());
        data.extend_from_slice(&1u16.to_le_bytes());
        data.extend_from_slice(&0u32.to_le_bytes());
        data.extend_from_slice(&0x8800_7070u32.to_le_bytes());
        data.extend_from_slice(&width.to_le_bytes());
        data.extend_from_slice(&height.to_le_bytes());
        data.extend_from_slice(&[0; 8]);
        data.extend_from_slice(&0x0064_7070u32.to_le_bytes());
        data.extend_from_slice(&width.to_le_bytes());
        data.extend_from_slice(&height.to_le_bytes());
        data.extend_from_slice(&[0; 4]);
        data.extend_from_slice(&16u16.to_le_bytes());
        data.extend_from_slice(&8u16.to_le_bytes());
        data.extend_from_slice(&64u32.to_le_bytes());
        data.extend_from_slice(&[0; 12]);
        for index in 0..32u8 {
            data.extend_from_slice(&[index, 255 - index, 64, 0x80]);
        }
        data
    }

    #[test]
    fn decodes_rgba_tiles_and_png() {
        let image = HgptImage::parse(&rgba_fixture()).unwrap();
        assert_eq!((image.width, image.height), (4, 8));
        assert_eq!(&image.rgba[..4], &[0, 255, 64, 255]);
        assert_eq!(&image.rgba[4..8], &[1, 254, 64, 255]);
        assert_eq!(&image.encode_png().unwrap()[..8], b"\x89PNG\r\n\x1a\n");
    }
}

File: tools/nge2-gltf/nge2_gltf/hgpt.py
from __future__ import annotations

from dataclasses import dataclass
from io import BytesIO

from PIL import Image

from .binary import Reader, align
from .errors import ParseError, UnsupportedFeature


@dataclass(frozen=True)
class HgptDivision:
    x: int
    y: int
    width: int
    height: int


@dataclass(frozen=True)
class HgptImage:
    width: int
    height: int
    format_code: int
    divisions: tuple[HgptDivision, ...]
    rgba: bytes

    @classmethod
    def parse(cls, data: bytes, *, resource: str = "texture.hpt") -> HgptImage:
        reader = Reader(data, resource)
        if bytes(reader.require(0, 4, "HGPT signature")) != b"HGPT":
            raise ParseError("missing HGPT signature", resource=resource, offset=0)
        pp_offset = reader.u16(4, "PP offset")
        extended = reader.u16(6, "extended-header flag")
        division_count = reader.u16(8, "division count")
        if pp_offset < 16:
            raise ParseError("PP offset precedes header", resource=resource, offset=4)
        if extended not in (0, 1):
            raise ParseError("extended-header flag must be 0 or 1", resource=resource, offset=6)
        divisions: list[HgptDivision] = []
        if extended:
            if reader.u16(16, "repeated division count") != division_count:
                raise ParseError("division counts do not match", resource=resource, offset=16)
            cursor = 28
            for _ in range(division_count):
                divisions.append(
                    HgptDivision(
                        reader.u16(cursor),
                        reader.u16(cursor + 2),
                        reader.u16(cursor + 4),
                        reader.u16(cursor + 6),
                    )
                )
                cursor += 8
            if cursor > pp_offset:
                raise ParseError(
                    "division table overlaps PP data", resource=resource, offset=cursor
                )

        pp_header = reader.u32(pp_offset, "PP header")
        if pp_header & 0xFFFF != 0x7070:
            raise ParseError("missing PP header", resource=resource, offset=pp_offset)
        format_code = pp_header >> 16
        formats = {0x14: (32, 1, 2), 0x13: (16, 1, 1), 0x8800: (4, 4, 1)}
        if format_code not in formats:
            raise UnsupportedFeature(
                f"HGPT pixel format 0x{format_code:04X}", resource=resource, offset=pp_offset
            )
        tile_width, bytes_per_unit, units_per_byte = formats[format_code]
        width = reader.u16(pp_offset + 4, "image width")
        height = reader.u16(pp_offset + 6, "image height")
        if width == 0 or height == 0 or width * height > 32_000_000:
            raise ParseError("invalid image dimensions", resource=resource, offset=pp_offset + 4)
        ppd_offset = pp_offset + 16
        ppd = reader.u32(ppd_offset, "PPD header")
        if ppd & 0x00FF_FFFF != 0x0064_7070 or ppd >> 24 != format_code & 0xFF:
            raise ParseError(
                "missing or mismatched PPD header", resource=resource, offset=ppd_offset
            )
        if reader.u16(ppd_offset + 4) != width or reader.u16(ppd_offset + 6) != height:
            raise ParseError(
                "PP and PPD dimensions differ", resource=resource, offset=ppd_offset + 4
            )
        storage_width = align(width, tile_width)
        storage_height = align(height, 8)
        unit_count = storage_width * storage_height
        data_size = unit_count * bytes_per_unit // units_per_byte
        pixel_offset = ppd_offset + 32
        tiled = reader.require(pixel_offset, data_size, "tiled pixels")
        palette: tuple[tuple[int, int, int, int], ...] = ()
        if format_code != 0x8800:
            palette_offset = pixel_offset + data_size
            if reader.u32(palette_offset, "PPC header") != 0x0063_7070:
                raise ParseError(
                    "missing PPC palette header", resource=resource, offset=palette_offset
                )
            count = reader.u16(palette_offset + 6, "palette block count") * 8
            expected = 16 if format_code == 0x14 else 256
            if count != expected:
                raise ParseError(
                    f"palette contains {count} colors, expected {expected}",
                    resource=resource,
                    offset=palette_offset + 6,
                )
            raw_palette = reader.require(palette_offset + 16, count * 4, "palette colors")
            palette = tuple(
                (raw_palette[i], raw_palette[i + 1], raw_palette[i + 2], _alpha(raw_palette[i + 3]))
                for i in range(0, len(raw_palette), 4)
            )

        rgba = bytearray(width * height * 4)
        for y in range(height):
            for x in range(width):
                tiled_index = _tile_index(x, y, storage_width, tile_width)
                if format_code == 0x14:
                    packed = tiled[tiled_index // 2]
                    color = palette[(packed >> 4) if tiled_index & 1 else (packed & 0x0F)]
                elif format_code == 0x13:
                    color = palette[tiled[tiled_index]]
                else:
                    source = tiled_index * 4
                    color = (
                        tiled[source],
                        tiled[source + 1],
                        tiled[source + 2],
                        _alpha(tiled[source + 3]),
                    )
                destination = (y * width + x) * 4
                rgba[destination : destination + 4] = bytes(color)
        return cls(width, height, format_code, tuple(divisions), bytes(rgba))

    @property
    def has_alpha(self) -> bool:
        return any(value != 255 for value in self.rgba[3::4])

    def encode_png(self) -> bytes:
        output = BytesIO()
        Image.frombytes("RGBA", (self.width, self.height), self.rgba).save(output, format="PNG")
        return output.getvalue()


def _tile_index(x: int, y: int, storage_width: int, tile_width: int) -> int:
    tile_size = tile_width * 8
    return (
        (y // 8) * tile_size * (storage_width // tile_width)
        + (x // tile_width) * tile_size
        + (y % 8) * tile_width
        + x % tile_width
    )


def _alpha(value: int) -> int:
    return min(value * 2, 255)
